app/ts/utils/uniswap.ts

```diff
@@ -0,0 +1,159 @@
+import { AbiCoder, getCreate2Address, keccak256, solidityPacked } from "ethers"
+import { EthereumAddress } from "./wire-types.js"
+import { Interface } from "ethers"
+import { UNISWAP_V2_FACTORY_ADDRESS, UNISWAP_V3_FACTORY_ADDRESS } from "./constants.js"
+import { addressString } from "./bigint.js"
+
+interface UniswapPools {
+	token0IsQuote: boolean
+	v2: EthereumAddress
+	v3: {
+		'100': EthereumAddress
+		'500': EthereumAddress
+		'3000': EthereumAddress
+		'10000': EthereumAddress
+	}
+}
+
+interface Multicall3Call {
+	target: string
+	allowFailure: boolean
+	callData: string
+}
+
+export function calculateUniswapPools(tokenA: EthereumAddress, tokenB: EthereumAddress): UniswapPools {
+	const [token0, token1] = tokenA < tokenB ? [addressString(tokenA), addressString(tokenB)] : [addressString(tokenB), addressString(tokenA)]
+	const abi = new AbiCoder()
+
+	const v2 = getCreate2Address(
+		addressString(UNISWAP_V2_FACTORY_ADDRESS),
+		keccak256(solidityPacked(['address', 'address'], [token0, token1])),
+		'0x96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f'
+	)
+
+	const v3 = {
+		'100': getCreate2Address(
+			addressString(UNISWAP_V3_FACTORY_ADDRESS),
+			keccak256(
+				abi.encode(['address', 'address', 'uint24'], [token0, token1, 100])
+			),
```

Comment:
These newlines don't seem to be adding value, recommend removing them.
```suggestion
			keccak256(abi.encode(['address', 'address', 'uint24'], [token0, token1, 100])),
```